libsql-remote/src/lib.rs

```diff
@@ -0,0 +1,290 @@
+    where
+        S: Serializer,
+    {
+        struct ColsRef<'a>(&'a [Col]);
+        impl<'a> Serialize for ColsRef<'a> {
+            fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+            where
+                S: Serializer,
+            {
+                let mut s = serializer.serialize_seq(Some(self.0.len()))?;
+                for col in self.0.iter() {
+                    s.serialize_element(&col.name)?;
+                }
+                s.end()
+            }
+        }
+
+        struct RowsRef<'a>(&'a VecDeque<Vec<Value>>);
+        impl<'a> Serialize for RowsRef<'a> {
+            fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+            where
+                S: Serializer,
+            {
+                let mut s = serializer.serialize_seq(Some(self.0.len()))?;
+                for row in self.0.iter() {
+                    s.serialize_element(&RowRef(row))?;
+                }
+                s.end()
+            }
+        }
+        struct RowRef<'a>(&'a [Value]);
+        impl<'a> Serialize for RowRef<'a> {
+            fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+            where
+                S: Serializer,
+            {
+                let mut s = serializer.serialize_seq(Some(self.0.len()))?;
+                for cell in self.0.iter() {
+                    match cell {
+                        Value::Null => s.serialize_element(&Value::Null),
+                        Value::Integer { value } => s.serialize_element(value),
+                        Value::Float { value } => s.serialize_element(value),
+                        Value::Text { value } => s.serialize_element(value),
+                        Value::Blob { value } => {
+                            let base64 = STANDARD_NO_PAD.encode(value);
+                            s.serialize_element(&json!({
+                                "base64": base64
+                            }))
+                        }
+                    }?;
+                }
+                s.end()
+            }
+        }
+
+        let mut s = serializer.serialize_struct("Rows", 2)?;
+        s.serialize_field("columns", &ColsRef(self.cols.deref()))?;
+
+        s.end()
+    }
+}
+
+impl Iterator for Rows {
+    type Item = Row;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let row = self.rows.pop_front()?;
+        Some(Row {
+            cols: self.cols.clone(),
+            inner: row,
+        })
+    }
+}
+
+impl Rows {
+    pub fn column_name(&self, idx: usize) -> Option<&str> {
+        self.cols
+            .get(idx)
+            .and_then(|c| c.name.as_ref())
+            .map(|s| s.as_str())
+    }
+
+    pub fn columns(&self) -> Columns {
+        self.cols.iter()
+    }
+}
+
+pub type Columns<'a> = std::slice::Iter<'a, Col>;
+
+pub type Cells<'a> = std::slice::Iter<'a, Value>;
```

Comment:
Why do we define these types instead of a `struct` for columns? These type aliases expose a lot of internal details that we probably don't want to expose right away.

Reply:
I was thinking about that initially, but there's so many useful traits (like Iterator/DoubleEndedIterator/ExactSizeIterator etc) that would benefit from being implemented, that I started to doubt how Columns or Cells are actually different to from the std::slice::Iter.